pkg/agent/cniserver/server.go

```diff
@@ -453,12 +453,12 @@ func (s *CNIServer) CmdAdd(ctx context.Context, request *cnipb.CniCmdRequest) (*
 		// Rollback to delete configurations once ADD is failure.
 		if !success {
 			if isInfraContainer {
-				klog.Warningf("CmdAdd for container %v failed, and try to rollback", cniConfig.ContainerId)
-				if _, err := s.CmdDel(ctx, request); err != nil {
-					klog.Warningf("Failed to rollback after CNI add failure: %v", err)
+				klog.InfoS("CmdAdd for container failed, trying to rollback", "container", cniConfig.ContainerId)
+				if _, err := s.cmdDel(ctx, cniConfig); err != nil {
```

Comment:
I agree that it is an improvement to directly use the existing "cniConfig" in cmdDel to avoid unnecessary call on `loadNetworkConfig`.

However, I didn't understand why the original code can append a duplicate PodCIDR range in the IPAM configuration. When we call function `loadNetworkConfig` in CmdDel/CmdAdd/CmdCheck, it has generated a new `CNIConfig` object, and the `cniConfig.NetworkConfig` is unmarshaled from the json string`request.CniArgs.NetworkConfiguration`. So even if we call CmdDel in CmdAdd rollback, the cniConfig passed to IPAM plugin from a new call with `loadNetworkConfig` is supposed to equal to the values existing in CmdAdd.

Reply:
The original code was calling `validateRequest` for the rollback (because `CmdDel` calls `validateRequest`), using the same `request` object. `validateRequest` actually _mutates_ the `request` but it is not obvious when looking at the code:

- `loadNetworkConfig` has the following assignment: https://github.com/antrea-io/antrea/blob/bbec9d04dd426bdcf1a7af0024ac93fdc61debda/pkg/agent/cniserver/server.go#L198 It means that mutating the contents of `cniConfig.CniCmdArgs` (a pointer to a protobuf struct) will mutate the request.
-  `updateLocalIPAMSubnet` has the following assignment: https://github.com/antrea-io/antrea/blob/bbec9d04dd426bdcf1a7af0024ac93fdc61debda/pkg/agent/cniserver/server.go#L275 `cniConfig.NetworkConfiguration` is actually the same as `cniConfig.CniCmdArgs.NetworkConfiguration` (it is a byte slice). So at this point we have mutated the request. The next time we call `validateRequestMessage` (which we no longer do with my patch), the network configuration of the request already includes the Node subnets in the IPAM section.

I confirmed all of this with my unit test.

Note that I don't know why we are mutating the request in the first place; I don't know if it's intentional, but I think we have been doing it for 4 years. Since using the existing `cniConfig` seemed like the right thing to do, that's what I did in this patch. We could consider a follow-up patch to clean up existing logic, but I didn't want to risk introducing a new bug.

Reply:
To clarify the statement above:

> `cniConfig.NetworkConfiguration` is actually the same as `cniConfig.CniCmdArgs.NetworkConfiguration`

This is because `CniCmdArgs` is an embedded field: https://github.com/antrea-io/antrea/blob/bbec9d04dd426bdcf1a7af0024ac93fdc61debda/pkg/agent/cniserver/server.go#L127-L136

Reply:
Got it, thanks for the clarification.